mcs/class/System/System.Net.Sockets/SocketAsyncResult.cs

```diff
@@ -301,5 +301,27 @@ public void Complete (Socket s, int total)
 			this.total = total;
 			Complete ();
 		}
+
+		void IThreadPoolWorkItem.ExecuteWorkItem()
+		{
+			switch (operation) {
+			case SocketOperation.Receive:
+				total = Socket.Receive_internal (socket.safe_handle, Buffer, Offset, Size, SockFlags, out error);
```

Comment:
given this can throw, we need to capture the exception so it can be later delivered once we complete.